File: src/main/java/clara/benchmark/drools/AccumRepeatedlyNotSelected.java
package clara.benchmark.drools;

import clara.benchmark.IBenchmark;
import clara.benchmark.beans.Customer;
import org.kie.api.KieBase;
import org.kie.api.runtime.KieSession;
import org.kie.api.runtime.rule.QueryResults;
import org.kie.api.runtime.rule.QueryResultsRow;

import java.util.ArrayList;
import java.util.List;

public class AccumRepeatedlyNotSelected implements IBenchmark<KieBase> {

    @Override
    public String getName() {
        return "drools.accum-repeatedly-not-selected";
    }

    @Override
    public KieBase getSession() throws Exception {
        return KnowledgeSupport.load("accum_repeatedly_not_selected.drl");
    }

    @Override
    public Iterable<?> getFacts() throws Exception {
        List<Customer> customers = new ArrayList<Customer>();
        for (int i = 1; i < 100; i++) {
            customers.add(new Customer(i, "CUSTOMER_NAME", true));
        }
        return customers;
    }

    @Override
    public void run(KieBase base, Iterable<Object> facts) throws Exception {

        KieSession session = base.newKieSession();

        try {
            for (Object fact : facts) {

                session.insert(fact);
                session.fireAllRules();

                QueryResults results = session.getQueryResults("get_discounts");

                // Iterate through results to ensure all work is done.
                for (QueryResultsRow result : results) {
                    result.get("$percent");
                }
            }
        } finally {
            session.dispose();
        }
    }
}

File: src/main/clojure/clara/benchmark/accum_repeatedly_not_selected.clj
(ns clara.benchmark.accum-repeatedly-not-selected
  "Added to test performance improvements in clara-rules issue 182"
  (:require [clara.rules :refer :all]
            [clara.rules.accumulators :as acc])
  (:import [clara.benchmark.beans Discount Customer]
           [java.util
            ArrayList
            List]))

(def min-id (acc/min (fn [^Customer c] (.getId c)) :returns-fact true))

(defrule vip-discount
  [?customer <- min-id :from [Customer]]
  =>
  (insert-all! (for [n (range 1000)] (Discount. 10))))

(defquery get-discounts
  []
  [Discount (= ?percent percent)])
